### DISCUSSION TITLE
How to set the clustering parameters to 0 mismatch in assemble?

### DISCUSSION THREAD

**Discussion by fookloy** [milaboratory/mixcr#1501] [Q&A] [answered]
[clustering-parameters](https://mixcr.com/mixcr/reference/mixcr-assemble/#clustering-parameters)

Hi guys,

I wanna do the assemble with clustering-parameters=0 mismatch by running

 `mixcr assemble  -OcloneClusteringParameters.searchParameters=oneMismatch   alignments.vdjca clones_m1.clns &` .

But it seems there have no command that could set the clustering-parameters to 0 mismatch.

Appreciate any suggestions, thanks!

**Comment by mizraelson** [comment]
If I understand correctly, having 0 mismatches means that cluster formation is not possible since no mismatches are permitted. In this scenario, you can disable it by using:

`-OcloneClusteringParameters=null`

**Reply by fookloy** [reply]
Thanks for your reply.

Since my data was sequenced using PCR-seq, 2*250 protocol, there must be many duplicate sequences in my data.

We first ran the command:
 
`mixcr align \
    --preset generic-amplicon \
    --species human \
    -OvParameters.geneFeatureToAlign="VRegionWithP" \
    -OvParameters.parameters.floatingLeftBound=false \
    -OjParameters.parameters.floatingRightBound=false \
    -OcParameters.parameters.floatingRightBound=false \
     HB-HV_L1_1.fq.gz \
     HB-HV_L1_2.fq.gz \
     HB-HV_L1.vdjca`

to get the .vdjca file, then i want to cluster (with the "assemble" command ) all the same sequences (0 mismatch) together with for the next steps , like the "fine allele" and "find shmtree", therefore, i am wondering will it stop doing the clustering step if we use the `-OcloneClusteringParameters=null` command?

**Reply by mizraelson** [reply]
I think you might be confusing clustering with clonotype assembly. Clustering is an algorithm that deals with PCR errors, while all identical sequences will be still assembled into clones regardless of clustering.

What region do your reads cover? I assume that 250+250 is not enough to cover the full VDJ region?

**Reply by fookloy** [reply]
Sorry for the mistake, clonotype assembling is what i meant.

So the clonotype assembly step will assemble identical sequences without mismatch right?

Our primer-F was designed in a conserved regions in FR1, while the primer-R was designed in CH1 close to the J region, the designed length is within 500 bp, around 460 bp.

**Reply by mizraelson** [reply, marked as answer]
I see, so I would recommend using the command bellow for a start:

```
mixcr analyze generic-amplicon \
    --species hsa \
    --rna \
    --floating-left-alignment-boundary \
    --floating-right-alignment-boundary C \
    --assemble-clonotypes-by "{CDR1Begin:FR4End}" \
      input_R1.fastq.gz \
      input_R2.fastq.gz \
      result
```

The command above will align and assemble clones by the sequence covered by your reads (starting from CDR1 to the end of FR4).
The resulting `clns` files can be used for `findAlleles` and `findShmTrees`.

**Reply by fookloy** [reply]
Really appreciate your help, i've tried this yesterday after your remind, and my progress is working good now~

But i am still confused that will `--assemble-clonotypes-by "{CDR1Begin:FR4End}"` do any correction or assemble my clonotype together with any mismatch, insertion or deletion after turning clustering parameters off using:  `-OcloneClusteringParameters=null`?

And here is one more question, are these two command different to each other? 

` --assemble-clonotypes-by "{CDR1Begin:FR4End}"`
`-OassemblingFeatures="VDJRegion"`

as i am concerned, my sequences contain most FR1 region to FR4End sequence, if i assemble them by CDR1 to FR4end, a part of FR1 will not count in, right? 

How about i enlarged the assemble region to VDJRegion, will this command assemble my clonotype concerning the full sequence?

**Reply by mizraelson** [reply]
So, I do not recommend using `-OcloneClusteringParameters=null` because it turns off the error-correction. This error-correction involves a sophisticated algorithm that evaluates the probability of error in each case and has proven to be very effective. If you plan to analyze alleles and somatic hypermutation (SHM) trees later, it is crucial to correct PCR errors so they won't be mistakenly counted as allelic variants and hypermutations. 

It's important to note that this feature does not indiscriminately merge all clones differing by a single mismatch. For instance, if you have two clones that differ by just one nucleotide, and both are supported by a significant number of reads, and in the majority of these reads the "mismatched" nucleotide is of high sequencing quality, the result will be two distinct clones. This nuanced approach ensures that genuine biological differences are maintained and accurately reflected in your analysis.

The parameter `--assemble-clonotypes-by` in `mixcr analyze` serves the same function as `-OassemblingFeatures` in `mixcr assemble`. You can't use `VDJRegion` to assemble clones because it requires full coverage of the FR1 region. Given that your primers are located in FR1, you cannot depend on the sequence covered by the primers, as any hypermutations present would not be detected. Therefore, it's preferable to assemble clones by the feature actually covered by the sequencing (not the primers). For example, if you know that all your primers end before the 10th nucleotide of FR1End, you could use the feature "{CDR1Begin(-10):FR4End}". This feature will start 10 nucleotides before CDR1.
